### PR TITLE
Fix issue: Wrong duration for DB transaction event  on ROR 7.1 #92

### PR DESCRIPTION
**Pull request by ibnjunaid**
*Issue #, if available:*

*Description of changes:*

X-Ray SDK expects time in seconds with up to nanosecond precision.
This change  make sure that the time is passed in seconds to the X-Ray for DB transactions. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/aws-xray-sdk/tc_time_conversion.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+require 'aws-xray-sdk/utils/math_utils'
+
+class TestGetTimeInSeconds < Minitest::Test
+    def convert_time_from_milliseconds
+        time_in_seconds = convert_time_in_seconds 1.723627099460531E12
+        assert_equal time_in_seconds 172362709.9460531
```
Comment (wangzlei) [inline test/aws-xray-sdk/tc_time_conversion.rb:6]: The expectation should be `1723627099.460531`

**Review by wangzlei — APPROVED** [review]
LGTM

**Comment by wangzlei** [comment]
The PR is to address https://github.com/aws/aws-xray-sdk-ruby/issues/92

File: lib/aws-xray-sdk/utils/math_utils.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    return 0 if num == 0
+    exp = Math.log10(num.abs).floor
+    num.negative? ? exp - 1 : exp
+end
+
+# X-Ray SDK expects time in seconds with nanosecond precision.
+def convert_time_in_seconds(time_float)
+    exponent = get_exponent(time_float)
+    division_factor = 10 ** (exponent - 9)
+    time_float / division_factor
```
Comment (jj22ee) [inline lib/aws-xray-sdk/utils/math_utils.rb:11]: Wouldn't this logic will break on `Saturday, November 20, 2286` when the input is (10^10), where the time will be converted to `Sunday, September 9, 2001`.

Perhaps there is something else we haven't thought of that can break this, maybe there is a better way of determining what value is being converted into seconds.

File: .github/workflows/continuous-build.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ jobs:
     strategy:
       matrix:
         os:
-          - macos-latest
+          - macos-13
```
Comment (jj22ee) [inline .github/workflows/continuous-build.yml:17]: Any reason to remove `macos-latest`? Why is it changed to `macos-13`, does having both help with testing all cases?

File: lib/aws-xray-sdk/utils/math_utils.rb (hunk 2 above)
Comment (jj22ee) [reply lib/aws-xray-sdk/utils/math_utils.rb:11]: Perhaps can the [rails version be checked](https://github.com/aws/aws-xray-sdk-ruby/issues/92#issuecomment-2291045563) to perform the division logic?

**Review by jj22ee — CHANGES_REQUESTED** [review]
Please take a look at my comments.

I believe you can check for rails version like:
```
if Rails::VERSION::MAJOR >= 7
  # Do the new thing
else
  # Do it the old way
end
```

File: lib/aws-xray-sdk/utils/math_utils.rb (hunk 2 above)
Comment (ibnjunaid) [reply lib/aws-xray-sdk/utils/math_utils.rb:11]: Sure, @jj22ee let me have a look.

File: .github/workflows/continuous-build.yml (hunk 3 above)
Comment (ibnjunaid) [reply .github/workflows/continuous-build.yml:17]: this was changed @wangzlei here to block the pipelines https://github.com/aws/aws-xray-sdk-ruby/commit/26ae7b08c1799a8fd44718f658390d7ab3b5425b

File: lib/aws-xray-sdk/utils/math_utils.rb (hunk 2 above)
Comment (ibnjunaid) [reply lib/aws-xray-sdk/utils/math_utils.rb:11]: On diving deep I observed that this issue is fixed in rails 7.2.

[https://github.com/rails/rails/pull/50767, https://github.com/rails/rails/pull/50493]

Do we want to conditionally handle for 7.1, something like this.

```
subsegment.start_time = (::Rails::VERSION::MAJOR == 7 and ::Rails::VERSION::MINOR == 1) ? convert_time_in_seconds(transaction.time.to_f) : transaction.time.to_f
          subsegment.sql = sql
          XRay.recorder.end_subsegment end_time: (::Rails::VERSION::MAJOR == 7 and ::Rails::VERSION::MINOR == 1) ? convert_time_in_seconds(transaction.end.to_f) : transaction.end.to_f
```

**Comment by jj22ee** [comment]
TY, can you update `convert_time_in_seconds` to be specifically for Rails 7.1, and update unit tests appropriately? Seems we only want to divide by `1000` only if Rails version is 7.1.

**Comment by ibnjunaid** [comment]
> convert_time_in_seconds

@jj22ee How do you suggest.

should we do something like this 

```ruby
          subsegment.start_time = (::Rails::VERSION::MAJOR == 7 and ::Rails::VERSION::MINOR == 1) ? convert_time_in_seconds(transaction.time.to_f) : transaction.time.to_f

````
or 
```ruby
   subsegment.start_time = (::Rails::VERSION::MAJOR == 7 and ::Rails::VERSION::MINOR == 1) ? transaction.time.to_f / 1000 : transaction.time.to_f
   ```

**Comment by ibnjunaid** [comment]
@jj22ee 


I am in favour of the first one because the changes are being back ported to some rails 7.1 version so I think its not a good idea to  just divide by 1000 without checking.

```Ruby
          subsegment.start_time = (::Rails::VERSION::MAJOR == 7 and ::Rails::VERSION::MINOR == 1) ? convert_time_in_seconds(transaction.time.to_f) : transaction.time.to_f
```

**Comment by jj22ee** [comment]
>Like has been back ported to 7-1-stable https://github.com/rails/rails/pull/50493#issuecomment-1872827961

1. This is saying the bug is back ported to 7.1, the [fix isn't backported from looking at this issue](https://github.com/rails/rails/pull/50767)
2. The function you are proposing will still have [the code smell](https://github.com/aws/aws-xray-sdk-ruby/pull/96#discussion_r1722118091) I've commented here

Can you provide some test results for Rails v7, v7.1, v7.2 for your change?

**Comment by jj22ee** [comment]
We can see here that [the fix](https://github.com/rails/rails/pull/50767) is applied in [7.1-stable branch here](https://github.com/rails/rails/blame/7-1-stable/activerecord/lib/active_record/runtime_registry.rb#L37-L38)

**Comment by ibnjunaid** [comment]
@jj22ee . The issue is still there for the available versions of 7.1.x.y of rails . 
Theres not any rails version with stable release it is just a branch in the repo as of now. 
https://rubygems.org/gems/rails/versions

**Review by jj22ee — APPROVED** [review]

**Pull request merged**